modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,469 @@
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            reusableBuff.putLong(addr + 1);
+            reusableBuff.flip();
+
+            while (reusableBuff.hasRemaining()) {
+                int bytesWritten = fileIo.write(reusableBuff);
+
+                if (bytesWritten <= 0)
+                    throw new IOException("Can not write data to file: " + idxFile.getAbsolutePath());
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets arbitrary file position.
+     *
+     * @param slot Position to set.
+     */
+    private void gotoSlot(long slot) {
+        try {
+            fileIo.position(slot * Entry.ENTRY_BYTES);
+        }
+        catch (Exception e) {
+            U.closeQuiet(fileIo);
+
+            throw new IgniteException("Failed to reset the index spill file, slot=" + slot, e);
+        }
+    }
+
+    /**
+     * Checks if the capacity of the hashtable is enough and extends it if needed.
+     */
+    private void ensureCapacity() {
+        if (entriesCnt > LOAD_FACTOR * cap) {
+            FileIO oldFileIo = fileIo;
+            File oldIdxFile = idxFile;
+
+            long oldSize = cap;
+
+            initNewIndexFile(oldSize * 2);
+
+            copyDataFromOldFile(oldFileIo, oldIdxFile, oldSize);
+        }
+    }
+
+    /**
+     * Copies data from the ols file to the extended new one.
+     *
+     * @param oldFile Old file channel.
+     * @param oldIdxFile Old index file.
+     * @param oldSize Old file size in slots.
+     */
+    private void copyDataFromOldFile(FileIO oldFile, File oldIdxFile, long oldSize) {
+        try {
+            entriesCnt = 0;
+
+            oldFile.position(0);
+
+            for (long i = 0; i < oldSize; i++) {
+                Entry e = readEntryFromIndexFile(null, oldFile);
+
+                if (!e.isRemoved() && !e.isEmpty())
+                    putEntryToFreeSlot(e.hashCode(), e.rowAddress());
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(fileIo);
+            U.closeQuiet(oldFile);
+
+            throw new IgniteException("Failed to extend hash index.", e);
+        }
+        finally {
+            U.closeQuiet(oldFile);
+            oldIdxFile.delete();
+        }
+    }
+
+    /**
+     * Initiates new file.
+     *
+     * @param cap New file capacity in entries (rows). Should be the power of 2.
+     */
+    private void initNewIndexFile(long cap) {
+        try {
+            assert cap > 0 && (cap & (cap - 1)) == 0; // Should be the positive power of 2.
+
+            this.cap = cap;
+
+            idxFile = new File(dir, spillFileName + "_idx_" + id++);
+
+            idxFile.deleteOnExit();
```

Comment:
This is synchronous call to add file to linked hash set.
Is it possible to add some tag to file header to we can detect if file was created with this HashIndex instance or not.
File from previous run can be safely deleted on fly with e.g. warn in log.

Reply:
I am not sure this is a performance problem. Creating a new index file is a quite rare event, so adding it to a sync set looks not expensive here.